### PR TITLE
Snow 1058245 SqlAlchemy 2.0 support

### PR DESCRIPTION
**Pull request by sfc-gh-mraba**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-1058245

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [x] I am adding compatibility with SQLAlchemy 2.0

3. Please describe how your code solves the related issue.

We adapted SnowflakeDialect compatibility with SQLAlchemy 2.0.
We added testing jobs with SQLAlchemy 2.0 installed.
We moved from tox to hatch.

### CODE REVIEW

**Comment by sfc-gh-astus** [comment]
I have read the CLA Document and I hereby sign the CLA

File: tests/conftest.py
Comment (sfc-gh-astus) [inline tests/conftest.py:159]: Is this needed?

File: src/snowflake/sqlalchemy/compat.py
Comment (sfc-gh-pjob) [inline src/snowflake/sqlalchemy/compat.py:18]: 1. Do I correctly understand it is sth like "args reducer for version < 2.0)? If yes, then is it possible to somehow include this info in the decorator name without making it too descriptive? The same question to `args_filler`.
2. Who is calling functions decorated with this decorator? I'm not sure if I understand why this decorator is needed.
3. Is this solution will work with keyword arguments? Does it have to?

File: src/snowflake/sqlalchemy/compat.py
Comment (sfc-gh-pjob) [inline src/snowflake/sqlalchemy/compat.py:34]: I don't understand who this warning is addressed to?

File: src/snowflake/sqlalchemy/compat.py
Comment (sfc-gh-mraba) [reply src/snowflake/sqlalchemy/compat.py:18]: Signature of methods for ormselectcompilestate has changed - 2 parameters on position 6 and 7 were removed. This resulted in arguments loss. It does not need to work with anything else than tuple - as annotated.

This will be removed when we drop support for 1.4

The methods are called internally by orm engine when performing joins.

I'm open to suggestion for more convenient name 👍

File: src/snowflake/sqlalchemy/compat.py
Comment (sfc-gh-mraba) [reply src/snowflake/sqlalchemy/compat.py:34]: This one was my own for testing.
This compatibility fix will have to be removed when 1.4 support will be dropped.

File: tests/conftest.py
Comment (sfc-gh-mraba) [reply tests/conftest.py:159]: Comments removed.

File: src/snowflake/sqlalchemy/compat.py
Comment (sfc-gh-pjob) [reply src/snowflake/sqlalchemy/compat.py:17]: Let's stay with current name.

**Review by sfc-gh-pjob — APPROVED** [review]

**Comment by antonrh** [comment]
🚀

**Comment by nitin-kakkar** [comment]
@sfc-gh-mraba - Thanks for adding 2.0 support. Do have any timeline of when this version will be released! 🥺

**Comment by nitin-kakkar** [comment]
> @sfc-gh-mraba - Thanks for adding 2.0 support. Do you have any timeline of when this version will be released? 🥺

Would really appreciate if you could provide some timeline. Thanks!!

**Comment by satniks** [comment]
It looks like implementation is completed and pull request is also approved. Any timelines for merging it and releasing ne version of snowflake-sqlalchemy?

**Comment by sjhewitt** [comment]
fwiw, I've been running this code in production for a few days and had no issues

**Comment by sfc-gh-dszmolka** [comment]
hey folks thank you for the interest; please head to the pinned issue https://github.com/snowflakedb/snowflake-sqlalchemy/issues/380 for timeline and updates - i'll post there as soon as there's any news or more fresh ETA

@sjhewitt appreciate your comment very much, good to hear the changes are running without major issues !

**Comment by nitin-kakkar** [comment]
@sfc-gh-mraba - Is there any update on timelines for merging and releasing new version of snowflake-sqlalchemy?

**Comment by sfc-gh-dszmolka** [comment]
@nitin-kakkar please head to the pinned issue https://github.com/snowflakedb/snowflake-sqlalchemy/issues/380 for timeline and updates - i'll post there as soon as there's any news or more fresh ETA

**Comment by mleot** [comment]
Is this PR going to be merged soon? I saw changes have been approved but not all the checks are complete.

**Comment by artem-sereda** [comment]
> Is this PR going to be merged soon? I saw changes have been approved but not all the checks are complete.

+1

**Comment by mleot** [comment]
It looks like two actions are stuck but are required before automatically merging. Maybe a maintainer would be able to manually start these two actions?

**Pull request merged**
